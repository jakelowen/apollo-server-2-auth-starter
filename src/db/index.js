const environment = process.env.NODE_ENV || 'development';
const knexTinyLogger = require('knex-tiny-logger').default;
const config = require('../../knexfile')[environment];

// eslint-disable-next-line import/order
const db = require('knex')(config);

const { DEBUG } = process.env;
if (DEBUG) knexTinyLogger(db);

module.exports = db;
