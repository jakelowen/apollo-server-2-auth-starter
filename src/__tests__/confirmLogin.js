/* eslint-disable no-undef */
import gql from 'graphql-tag';
import { createTestClient } from 'apollo-server-testing';
import jwt from 'jsonwebtoken';
import { constructTestServer, makeTestContext } from './__utils';
import db from '../db';

require('dotenv').config();

const CONFIRM_LOGIN_MUTATION = gql`
    mutation confirmLogin($token: String!){
      confirmLogin(token: $token) {
            code
            success
            message
            user {
              id
            }
        }
    }
`;

beforeAll(async () => db.migrate.latest());
beforeEach(async () => Promise.all([db.raw('TRUNCATE TABLE users CASCADE')]));
afterAll(async () => db.destroy());

describe('ConfirmLogin', () => {
  test('Happy Path', async () => {
    // insert record
    const loginToken = 'look at me I am a test string';
    const email = 'fake123@test.com';
    const id = '0b986756-e6e3-4bec-bb3c-fe5ce28d0751';
    await db.table('users').insert({
      id, name: 'fake dude', email, login_token: loginToken, login_token_expiry: db.raw('now() + \'24 HOUR\'::INTERVAL'),
    });
    await db('permission').insert([
      { user_id: id, role: 'foo' },
      { user_id: id, role: 'bar' },
    ]);
    const cookie = jest.fn();

    const { server } = constructTestServer({
      context: await makeTestContext(null, { res: { cookie } }),
    });

    const { mutate } = createTestClient(server);
    const res = await mutate({
      mutation: CONFIRM_LOGIN_MUTATION,
      variables: { token: loginToken },
    });

    // verify cookie is set
    expect(cookie).toHaveBeenCalled();

    // verify database record removes login_token
    const dbUser = await db('users').where({ email }).first();
    expect(dbUser.loginToken).toBe(undefined);
    expect(dbUser.login_token_expiry).toBe(null);

    // grab token from mock function
    const authTokenCalls = cookie.mock.calls[0];
    expect(authTokenCalls[0]).toBe('token');
    expect(authTokenCalls[2]).toEqual({
      httpOnly: true,
      maxAge: 1000 * 60 * 60 * 24 * 365,
    });
    const authToken = authTokenCalls[1];
    // 1 verify is legit token
    const decodedAuthToken = jwt.verify(authToken, process.env.TOKEN_SECRET);


    // 2 verify permission roles match
    // expect(decodedAuthToken.roles).toContain('applicant');
    // expect(decodedAuthToken.roles).toContain('foo');
    // expect(decodedAuthToken.roles).toContain('bar');
    // 3 verify user id
    expect(decodedAuthToken.userId).toBe(id);

    // all as snapshot?
    expect(res).toMatchSnapshot();
  });


  test('throws error if logged in', async () => {
    const { server } = constructTestServer({
      context: await makeTestContext(null, { user: { id: 1, email: 'a@a.a' } }),
    });
    const { mutate } = createTestClient(server);
    const res = await mutate({ mutation: CONFIRM_LOGIN_MUTATION, variables: { token: 'nonsense' } });
    expect(res).toMatchSnapshot();
  });

  test('response if no matching user', async () => {
    const { server } = constructTestServer({
      context: await makeTestContext(),
    });

    const { mutate } = createTestClient(server);
    const res = await mutate({
      mutation: CONFIRM_LOGIN_MUTATION,
      variables: { token: 'nonsense' },
    });
    expect(res).toMatchSnapshot();
  });
});
