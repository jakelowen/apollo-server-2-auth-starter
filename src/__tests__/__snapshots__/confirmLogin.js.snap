// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmLogin Happy Path 1`] = `
Object {
  "data": Object {
    "confirmLogin": Object {
      "code": "ok",
      "message": "You have successfully logged in. You may close this window and return to your previous tab to continue!",
      "success": true,
      "user": Object {
        "id": "0b986756-e6e3-4bec-bb3c-fe5ce28d0751",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`ConfirmLogin response if no matching user 1`] = `
Object {
  "data": Object {
    "confirmLogin": Object {
      "code": "noUser",
      "message": "This token is either invalid or expired!",
      "success": false,
      "user": null,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`ConfirmLogin throws error if logged in 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You must not be logged in to perform this action.],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
